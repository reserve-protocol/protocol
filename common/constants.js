const { BigNumber } = require("ethers");

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

const ONE_ETH= BigNumber.from("1000000000000000000");

const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1);
const MAX_UINT16 = (2**16) - 1;

const SCALE_FACTOR = 1e18;
const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString());

module.exports = {
    ZERO_ADDRESS,
    MAX_UINT256,
    MAX_UINT16,
    SCALE_FACTOR,
    BN_SCALE_FACTOR,
    ONE_ETH
};
