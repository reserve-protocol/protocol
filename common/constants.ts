import { ethers } from 'hardhat'
import { BigNumber } from 'ethers'

export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
export const ONE_ADDRESS = '0x0000000000000000000000000000000000000001'

export const ONE_ETH = BigNumber.from('1000000000000000000')

export const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1)
export const MAX_UINT192 = BigNumber.from(2).pow(192).sub(1)
export const MAX_UINT96 = BigNumber.from(2).pow(96).sub(1)
export const MAX_UINT32 = 2 ** 32 - 1
export const MAX_UINT16 = 2 ** 16 - 1

export const SCALE_DECIMALS = 18
export const SCALE_FACTOR = 10 ** SCALE_DECIMALS
export const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString())

// @dev Must match `IAsset.CollateralStatus`.
export enum CollateralStatus {
  SOUND,
  IFFY,
  DISABLED,
}

// @dev Must match `Governance.ProposalState`.
export enum ProposalState {
  Pending,
  Active,
  Canceled,
  Defeated,
  Succeeded,
  Queued,
  Expired,
  Executed,
}

// @dev Must match `Fixed.RoundingApproach`.
export enum RoundingMode {
  FLOOR,
  ROUND,
  CEIL,
}

// @dev Must match `GnosisTrade.TradeStatus`.
export enum TradeStatus {
  NOT_STARTED,
  OPEN,
  CLOSED,
}

export const FURNACE_DEST = '0x0000000000000000000000000000000000000001'
export const STRSR_DEST = '0x0000000000000000000000000000000000000002'

export const QUEUE_START = '0x0000000000000000000000000000000000000000000000000000000000000001'

// Auth roles
export const OWNER = ethers.utils.formatBytes32String('OWNER')
export const SHORT_FREEZER = ethers.utils.formatBytes32String('SHORT_FREEZER')
export const LONG_FREEZER = ethers.utils.formatBytes32String('LONG_FREEZER')
export const PAUSER = ethers.utils.formatBytes32String('PAUSER')
