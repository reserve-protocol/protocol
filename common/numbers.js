const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}