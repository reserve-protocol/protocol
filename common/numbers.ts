import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
