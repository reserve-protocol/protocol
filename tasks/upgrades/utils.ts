import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { IComponents } from '../../common/configuration'
import { IDeployments } from '../../scripts/deployment/common'
import { isValidContract } from '../../common/blockchain-utils'

export const validateDeployments = async (
  hre: HardhatRuntimeEnvironment,
  deployments: IDeployments,
  version: string
) => {
  // Check trading lib defined
  if (!deployments.tradingLib) {
    throw new Error(
      `Missing deployed TradingLib for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.tradingLib))) {
    throw new Error(`TradingLib contract not found in network ${hre.network.name}`)
  }

  // Check Main implementation is defined
  if (!deployments.implementations.main) {
    throw new Error(
      `Missing deployed Main implementation for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.implementations.main))) {
    throw new Error(`Main contract not found in network ${hre.network.name}`)
  }

  // Check all componet implementations are defined
  const componentsKeys: string[] = [
    'backingManager',
    'basketHandler',
    'broker',
    'distributor',
    'furnace',
    'rTokenTrader',
    'rsrTrader',
    'rToken',
    'stRSR',
  ]

  componentsKeys.forEach(async (keystr) => {
    if (!deployments.implementations.components[keystr as keyof IComponents]) {
      throw new Error(
        `Missing deployed ${keystr} implementation for version ${version} in network ${hre.network.name}`
      )
    } else if (
      !(await isValidContract(
        hre,
        deployments.implementations.components[keystr as keyof IComponents]
      ))
    ) {
      throw new Error(
        `Implementation ${keystr} contract not found for version ${version} in network ${hre.network.name}`
      )
    }
  })
}
