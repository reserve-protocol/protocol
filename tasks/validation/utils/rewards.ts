import { TradeKind } from '#/common/constants'
import { advanceBlocks, advanceTime } from '#/utils/time'
import { IRewardable } from '@typechain/IRewardable'
import { formatEther } from 'ethers/lib/utils'
import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { runBatchTrade } from './trades'

const claimRewards = async (claimer: IRewardable) => {
  const resp = await claimer.claimRewards()
  const r = await resp.wait()
  const rewards = []
  for (const event of r.events!) {
    if (event.event == 'RewardsClaimed' && event.args!.amount.gt(0)) {
      rewards.push(event.args!.erc20)
    }
  }
  return rewards
}

// Expects RToken was transferred into RSRTrader beforehand
export const processRevenue = async (hre: HardhatRuntimeEnvironment, rtokenAddress: string) => {
  console.log(`\n* * * * * Claiming RSR rewards...`)
  const rToken = await hre.ethers.getContractAt('RTokenP1', rtokenAddress)
  const main = await hre.ethers.getContractAt('IMain', await rToken.main())
  const assetRegistry = await hre.ethers.getContractAt(
    'AssetRegistryP1',
    await main.assetRegistry()
  )
  const backingManager = await hre.ethers.getContractAt(
    'BackingManagerP1',
    await main.backingManager()
  )
  const rsrTrader = await hre.ethers.getContractAt('RevenueTraderP1', await main.rsrTrader())
  const rsr = await hre.ethers.getContractAt('ERC20Mock', await main.rsr())
  const strsr = await hre.ethers.getContractAt('StRSRP1', await main.stRSR())
  const rsrRatePre = await strsr.exchangeRate()

  const [rewards, assets] = await assetRegistry.getRegistry()
  let successCount = 0
  for (let i = 0; i < rewards.length; i++) {
    try {
      await backingManager.claimRewardsSingle(rewards[i])
      successCount++
    } catch (e) {
      console.log(
        `❌ failed to claim rewards for asset ${assets[i]} - review, may be a false positive`
      )
    }
  }
  const emoji = successCount == rewards.length ? '✅' : '❌'
  console.log(
    `${emoji} claimRewardsSingle() was successful for ${successCount}/${rewards.length} assets`
  )
  // await claimRewards(backingManager)

  await backingManager.forwardRevenue(rewards)

  await rsrTrader.manageTokens([rToken.address], [TradeKind.BATCH_AUCTION])
  await runBatchTrade(hre, rsrTrader, rToken.address, false)

  await strsr.payoutRewards()
  await advanceBlocks(hre, 100)
  await advanceTime(hre, 1200)
  await strsr.payoutRewards()

  const rsrRatePost = await strsr.exchangeRate()
  if (!rsrRatePost.gt(rsrRatePre)) {
    throw new Error(
      `stRSR rate should have increased. pre: ${formatEther(rsrRatePre)}   post ${formatEther(
        rsrRatePost
      )}`
    )
  }

  console.log('Successfully claimed and distributed RSR rewards')
}
