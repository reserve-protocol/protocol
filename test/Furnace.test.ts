import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { expect } from 'chai'
import { BigNumber, ContractFactory, Wallet } from 'ethers'
import hre, { ethers, upgrades, waffle } from 'hardhat'
import { IConfig, MAX_PERIOD, MAX_RATIO } from '../common/configuration'
import { bn, fp } from '../common/numbers'
import { issueMany } from './utils/issue'
import {
  CTokenMock,
  ERC20Mock,
  FacadeRead,
  StaticATokenMock,
  TestIFurnace,
  TestIMain,
  TestIRToken,
  USDCMock,
} from '../typechain'
import { advanceBlocks, advanceTime } from './utils/time'
import { Collateral, defaultFixture, Implementation, IMPLEMENTATION } from './fixtures'
import { makeDecayFn } from './utils/rewards'
import snapshotGasCost from './utils/snapshotGasCost'
import { cartesianProduct } from './utils/cases'
import { ZERO_ADDRESS } from '../common/constants'
import { useEnv } from '#/utils/env'

const createFixtureLoader = waffle.createFixtureLoader

const describeGas =
  IMPLEMENTATION == Implementation.P1 && useEnv('REPORT_GAS') ? describe : describe.skip

describe(`FurnaceP${IMPLEMENTATION} contract`, () => {
  let owner: SignerWithAddress
  let addr1: SignerWithAddress
  let addr2: SignerWithAddress

  // Contracts
  let main: TestIMain
  let furnace: TestIFurnace
  let rToken: TestIRToken
  let facade: FacadeRead
  let basket: Collateral[]

  // Config
  let config: IConfig

  // Tokens/Assets
  let token0: ERC20Mock
  let token1: ERC20Mock
  let token2: StaticATokenMock
  let token3: CTokenMock

  let collateral0: Collateral
  let collateral1: Collateral
  let collateral2: Collateral
  let collateral3: Collateral

  let initialBal: BigNumber

  let loadFixture: ReturnType<typeof createFixtureLoader>
  let wallet: Wallet

  // Implementation-agnostic interface for deploying the Furnace
  const deployNewFurnace = async (): Promise<TestIFurnace> => {
    if (IMPLEMENTATION == Implementation.P0) {
      const FurnaceFactory: ContractFactory = await ethers.getContractFactory('FurnaceP0')
      return <TestIFurnace>await FurnaceFactory.deploy()
    } else if (IMPLEMENTATION == Implementation.P1) {
      const FurnaceFactory: ContractFactory = await ethers.getContractFactory('FurnaceP1')
      return <TestIFurnace>await upgrades.deployProxy(FurnaceFactory, [], {
        kind: 'uups',
      })
    } else {
      throw new Error('PROTO_IMPL must be set to either `0` or `1`')
    }
  }

  before('create fixture loader', async () => {
    ;[wallet] = (await ethers.getSigners()) as unknown as Wallet[]
    loadFixture = createFixtureLoader([wallet])
  })

  beforeEach(async () => {
    ;[owner, addr1, addr2] = await ethers.getSigners()

    // Deploy fixture
    ;({ main, basket, rToken, furnace, config } = await loadFixture(defaultFixture))

    // Setup issuance of RTokens for users
    initialBal = bn('100e18')

    // Get assets and tokens
    ;[collateral0, collateral1, collateral2, collateral3] = basket

    token0 = <ERC20Mock>await ethers.getContractAt('ERC20Mock', await collateral0.erc20())
    token1 = <USDCMock>await ethers.getContractAt('USDCMock', await collateral1.erc20())
    token2 = <StaticATokenMock>(
      await ethers.getContractAt('StaticATokenMock', await collateral2.erc20())
    )
    token3 = <CTokenMock>await ethers.getContractAt('CTokenMock', await collateral3.erc20())

    // Mint Tokens
    await token0.connect(owner).mint(addr1.address, initialBal)
    await token1.connect(owner).mint(addr1.address, initialBal)
    await token2.connect(owner).mint(addr1.address, initialBal)
    await token3.connect(owner).mint(addr1.address, initialBal)

    await token0.connect(owner).mint(addr2.address, initialBal)
    await token1.connect(owner).mint(addr2.address, initialBal)
    await token2.connect(owner).mint(addr2.address, initialBal)
    await token3.connect(owner).mint(addr2.address, initialBal)
  })

  describe('Deployment #fast', () => {
    it('Deployment should setup Furnace correctly', async () => {
      expect(await furnace.period()).to.equal(config.rewardPeriod)
      expect(await furnace.ratio()).to.equal(config.rewardRatio)
      expect(await furnace.lastPayout()).to.be.gt(0) // A timestamp is set
      expect(await furnace.main()).to.equal(main.address)
    })

    it('Deployment does not accept empty period', async () => {
      const newConfig = JSON.parse(JSON.stringify(config))
      newConfig.rewardPeriod = bn('0')
      const newFurnace: TestIFurnace = <TestIFurnace>await deployNewFurnace()
      await expect(
        newFurnace.init(main.address, newConfig.rewardPeriod, newConfig.rewardRatio)
      ).to.be.revertedWith('invalid period')
    })

    // Applies to all components - used here as an example
    it('Deployment does not accept invalid main address', async () => {
      const newFurnace: TestIFurnace = <TestIFurnace>await deployNewFurnace()
      await expect(
        newFurnace.init(ZERO_ADDRESS, config.rewardPeriod, config.rewardRatio)
      ).to.be.revertedWith('main is zero address')
    })
  })

  describe('Configuration / State #fast', () => {
    it('Should allow to update period correctly if Owner and perform validations', async () => {
      // Setup a new value
      const newRewardPeriod: BigNumber = bn('100000')

      await expect(furnace.connect(owner).setPeriod(newRewardPeriod))
        .to.emit(furnace, 'PeriodSet')
        .withArgs(config.rewardPeriod, newRewardPeriod)

      expect(await furnace.period()).to.equal(newRewardPeriod)

      // Try to update again if not owner
      await expect(furnace.connect(addr1).setPeriod(bn('500'))).to.be.revertedWith(
        'governance only'
      )

      // Cannot update with period zero
      await expect(furnace.connect(owner).setPeriod(bn('0'))).to.be.revertedWith('invalid period')

      // Cannot update with period > max
      await expect(furnace.connect(owner).setPeriod(MAX_PERIOD + 1)).to.be.revertedWith(
        'invalid period'
      )
    })

    it('Should allow to update ratio correctly if Owner and perform validations', async () => {
      // Setup a new value
      const newRatio: BigNumber = bn('100000')

      await expect(furnace.connect(owner).setRatio(newRatio))
        .to.emit(furnace, 'RatioSet')
        .withArgs(config.rewardRatio, newRatio)

      expect(await furnace.ratio()).to.equal(newRatio)

      // Try to update again if not owner
      await expect(furnace.connect(addr1).setRatio(bn('0'))).to.be.revertedWith('governance only')

      // Cannot update with ratio > max
      await expect(furnace.connect(owner).setRatio(MAX_RATIO.add(1))).to.be.revertedWith(
        'invalid ratio'
      )
    })
  })

  describe('Do Melt #fast', () => {
    beforeEach(async () => {
      // Approvals for issuance
      await token0.connect(addr1).approve(rToken.address, initialBal)
      await token1.connect(addr1).approve(rToken.address, initialBal)
      await token2.connect(addr1).approve(rToken.address, initialBal)
      await token3.connect(addr1).approve(rToken.address, initialBal)

      await token0.connect(addr2).approve(rToken.address, initialBal)
      await token1.connect(addr2).approve(rToken.address, initialBal)
      await token2.connect(addr2).approve(rToken.address, initialBal)
      await token3.connect(addr2).approve(rToken.address, initialBal)

      // Issue tokens
      const issueAmount: BigNumber = bn('100e18')
      await rToken.connect(addr1).issue(issueAmount)
      await rToken.connect(addr2).issue(issueAmount)
    })

    it('Should not melt if paused', async () => {
      await main.connect(owner).pause()
      await expect(furnace.connect(addr1).melt()).to.be.revertedWith('paused or frozen')
    })

    it('Should not melt if frozen', async () => {
      await main.connect(owner).freezeShort()
      await expect(furnace.connect(addr1).melt()).to.be.revertedWith('paused or frozen')
    })

    it('Should not melt any funds in the initial block', async () => {
      const hndAmt: BigNumber = bn('2e18')

      // Set automine to false for multiple transactions in one block
      await hre.network.provider.send('evm_setAutomine', [false])

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Melt
      await furnace.connect(addr1).melt()

      // Mine block
      await hre.network.provider.send('evm_mine', [])

      // Check melt was not processed
      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Set automine to true again
      await hre.network.provider.send('evm_setAutomine', [true])
    })

    it('Should not melt if no funds available', async () => {
      // Set time period
      const period: number = 60 * 60 * 24 // 1 day
      await furnace.connect(owner).setPeriod(period)

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal)
      expect(await rToken.balanceOf(furnace.address)).to.equal(0)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      // Melt
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      // Check nothing changed
      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal)
      expect(await rToken.balanceOf(furnace.address)).to.equal(0)
    })

    it('Should melt 0 for first period, even if funds available', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance one period
      await advanceTime(period + 1)

      // Melt
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      // Another call to melt should also have no impact
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)
    })

    it('Should allow melt - one period', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 1) // 1 period

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(hndAmt.sub(expAmt))

      // Another call to melt should have no impact
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt)
    })

    it('Should allow melt - two periods, all at once', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(2 * period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 2) // 2 periods

      await expect(furnace.melt()).to.emit(rToken, 'Melted').withArgs(hndAmt.sub(expAmt))

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt)
    })

    it('Should allow melt - two periods, one at a time', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt1 = decayFn(hndAmt, 1) // 1 period

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(hndAmt.sub(expAmt1))

      // Advance to the end to withdraw full amount
      await advanceTime(period + 1)

      const expAmt2 = decayFn(hndAmt, 2) // 2 periods

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(bn(expAmt1).sub(expAmt2))

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt2)
    })
  })

  describe('Extreme Bounds', () => {
    const applyParameters = async (
      period: BigNumber,
      ratio: BigNumber,
      bal: BigNumber
    ): Promise<TestIFurnace> => {
      // Deploy fixture
      ;({ main, rToken, facade, furnace } = await loadFixture(defaultFixture))

      await furnace.connect(owner).setPeriod(period)
      await furnace.connect(owner).setRatio(ratio)

      const max256 = bn(2).pow(256).sub(1)
      await token0.connect(owner).mint(addr1.address, max256)
      await token1.connect(owner).mint(addr1.address, max256)
      await token2.connect(owner).mint(addr1.address, max256)
      await token3.connect(owner).mint(addr1.address, max256)
      await token0.connect(addr1).approve(rToken.address, max256)
      await token1.connect(addr1).approve(rToken.address, max256)
      await token2.connect(addr1).approve(rToken.address, max256)
      await token3.connect(addr1).approve(rToken.address, max256)

      // Issue and send tokens to furnace
      if (bal.gt(bn('0'))) {
        await issueMany(facade, rToken, bal, addr1)
      }

      // Charge battery
      await rToken.connect(owner).setRedemptionRateFloor(bal)
      await advanceBlocks(300)

      return furnace
    }

    it('Should not revert at extremes', async () => {
      // max: 1 year
      const periods = [bn(MAX_PERIOD), bn('1'), bn('604800')]

      const ratios = [fp('1'), fp('0'), fp('0.02284')]

      const bals = [fp('1e18'), fp('0'), bn('1e9')]

      const cases = cartesianProduct(periods, ratios, bals)
      for (let i = 0; i < cases.length; i++) {
        const args: BigNumber[] = cases[i]
        const period = args[0]
        const ratio = args[1]
        const bal = args[2]

        const newFurnace: TestIFurnace = <TestIFurnace>await applyParameters(period, ratio, bal)

        // Should melt after 1 period
        await advanceTime(period.add(1).toString())
        await newFurnace.melt()

        // Should melt after 1000 periods
        await advanceTime(period.mul(1000).add(1).toString())
        await newFurnace.melt()
      }
    })
  })

  describeGas('Gas Reporting', () => {
    beforeEach(async () => {
      // Approvals for issuance
      await token0.connect(addr1).approve(rToken.address, initialBal)
      await token1.connect(addr1).approve(rToken.address, initialBal)
      await token2.connect(addr1).approve(rToken.address, initialBal)
      await token3.connect(addr1).approve(rToken.address, initialBal)

      await token0.connect(addr2).approve(rToken.address, initialBal)
      await token1.connect(addr2).approve(rToken.address, initialBal)
      await token2.connect(addr2).approve(rToken.address, initialBal)
      await token3.connect(addr2).approve(rToken.address, initialBal)

      // Issue tokens
      const issueAmount: BigNumber = bn('100e18')
      await rToken.connect(addr1).issue(issueAmount)
      await rToken.connect(addr2).issue(issueAmount)

      // Advance blocks to fill battery
      await advanceBlocks(300)
    })

    it('Melt - One period ', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Call with no impact
      await snapshotGasCost(furnace.connect(addr1).melt())

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await snapshotGasCost(furnace.connect(addr1).melt())

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 1) // 1 period

      // Melt
      await snapshotGasCost(furnace.connect(addr1).melt())

      // Another call to melt with no impact
      await snapshotGasCost(furnace.connect(addr1).melt())

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt)
    })

    it('Melt - Many periods, all at once', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await snapshotGasCost(furnace.connect(addr1).melt())
      // Advance to the end to melt full amount
      await advanceTime(10 * period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 10) // 10 periods

      await snapshotGasCost(furnace.connect(addr1).melt())

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.be.closeTo(expAmt, 15)
    })

    it('Melt - Many periods, one after the other', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await snapshotGasCost(furnace.connect(addr1).melt())

      // Melt 10 periods
      for (let i = 1; i <= 10; i++) {
        await advanceTime(period + 1)
        await snapshotGasCost(furnace.connect(addr1).melt())
      }

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 10) // 10 periods

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.be.closeTo(expAmt, 15)
    })
  })
})
