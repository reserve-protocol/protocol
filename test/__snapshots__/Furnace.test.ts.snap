// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, all at once 1`] = `81334`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, all at once 2`] = `85581`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 1`] = `81334`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 2`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 3`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 4`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 5`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 6`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 7`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 8`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 9`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 10`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 11`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  1`] = `43192`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  2`] = `81334`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  3`] = `83415`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  4`] = `43192`;
